web/src/app/dialogs/new-inspection/components/file-upload.component.ts

```diff
@@ -0,0 +1,194 @@
+
+  /**
+   * The status of upload for this file form.
+   */
+  uploadStatus = input(UploadStatus.Waiting);
+
+  /**
+   * The state if currently selected file is uploaded or not.
+   */
+  uploaded = signal(true);
+
+  fileDraggingOverArea = signal(false);
+
+  /**
+   * The ratio of file size completed upload.
+   */
+  uploadRatio = signal(0);
+
+  /**
+   * The filename uploaded or will be uploaded on this field.
+   * This state directly hold by FileUploadComponent and not used except for users to know which they uploaded.
+   */
+  filename = signal('');
+
+  @ViewChild('fileInput')
+  fileInput!: ElementRef<HTMLInputElement>;
+
+  selectedFile: File | null = null;
+
+  private snackBar = inject(MatSnackBar);
+
+  private uploader = inject(FILE_UPLOADER);
+
+  /**
+   * Event handler of clicking the drop area.
+   */
+  onClickFileDialogOpen() {
+    this.fileInput.nativeElement.click();
+  }
+
+  /**
+   * Eventhandler for change event of the hidden file input opened the file dialog.
+   */
+  onSelectedFileChangedFromDialog() {
+    this.processReceivedFileInfo(
+      this.fileListToArray(this.fileInput.nativeElement.files),
+    );
+  }
+
+  /**
+   * Eventhandler for dragenter of the dropping area.
+   */
+  onFileDragEnter(e: DragEvent) {
+    this.fileDraggingOverArea.set(true);
+    e.preventDefault();
+  }
+
+  /**
+   * Eventhandler for dragleave of the dropping area.
+   */
+  onFileDragLeave() {
+    this.fileDraggingOverArea.set(false);
+  }
+
+  /**
+   * Eventhandler for dragover of the dropping area.
+   */
+  onFileDragOver(e: DragEvent) {
+    e.preventDefault(); // needs preventDefault() in dragover and dragenter not to open the file directly with the browser page.
+  }
+
+  /**
+   * Eventhandler for drop of the dropping area.
+   */
+  onFileDrop(e: DragEvent) {
+    e.preventDefault();
+    e.stopImmediatePropagation();
+    this.fileDraggingOverArea.set(false);
+    this.processReceivedFileInfo(this.fileListToArray(e.dataTransfer?.files));
+  }
+
+  /**
+   * Eventhandler for the upload button.
+   */
+  onClickUploadButton() {
+    if (this.selectedFile === null) {
+      return;
+    }
+    this.uploader
+      .upload(this.uploadToken(), this.selectedFile)
+      .subscribe((status) => {
+        this.uploadRatio.set(status.completeRatio);
+        if (status.done) {
+          this.uploaded.set(true);
+        }
+      });
+  }
+
+  processReceivedFileInfo(files: File[]) {
+    if (files.length > 1) {
+      this.snackBar.open('2 or more files are specified at once.');
+    }
+    const file = files[0];
+    this.filename.set(file.name);
+    this.uploaded.set(false);
+    this.selectedFile = file;
+  }
+
+  /**
+   * Convert FileList types to the list of File.
+   * Returns an empty array when the input is null or undefined.
+   */
+  private fileListToArray(files?: FileList | null): File[] {
+    if (files === undefined || files === null) {
+      return [];
+    }
+    const arrayOfFiles: File[] = [];
```

Comment:
maybe we can simply `return Array.from(files)` after null/undefined check?

Reply:
Nice suggestion 😄 
I rewrote them with `Array.from(files ?? [])`